specs/experimental/cannon-fault-proof-vm-mt.md

```diff
@@ -0,0 +1,511 @@
+   For details, see the [“Thread Stack Hashing” section.](#thread-stack-hashing)
+1. `nextThreadID` - 32-bit value defining the id to assign to the next thread that is created.
+
+The state is represented by packing the above fields, in order, into a 172-byte buffer.
+
+### State Hash
+
+The state hash is computed by hashing the 172-byte state buffer with the Keccak256 hash function
+and then setting the high-order byte to the respective VM status.
+
+The VM status can be derived from the state's `exited` and `exitCode` fields.
+
+```rs
+enum VmStatus {
+    Valid = 0,
+    Invalid = 1,
+    Panic = 2,
+    Unfinished = 3,
+}
+
+fn vm_status(exit_code: u8, exited: bool) -> u8 {
+    if exited {
+        match exit_code {
+            0 => VmStatus::Valid,
+            1 => VmStatus::Invalid,
+            _ => VmStatus::Panic,
+        }
+    } else {
+        VmStatus::Unfinished
+    }
+}
+```
+
+### Thread State
+
+The state of a single thread is tracked and represented by a thread state object consisting of the following fields:
+
+1. `threadID` - 32-bit unique thread identifier.
+1. `exitCode` - 8-bit exit code.
+1. `exited` - 8-bit boolean value indicating whether the thread has exited.
+1. `futexAddr` - 32-bit address set via a futex syscall indicating that this thread is waiting on a value change
+    at this address.
+1. `futexVal` - 32-bit value representing the memory contents at `futexAddr` when this thread began waiting.
+1. `futexTimeoutStep` - 64-bit value representing the future `step` at which the futex wait will time out.  Set to the
+   max uint64 value (-1) if no timeout is active.
+1. `pc` - 32-bit program counter.
+1. `nextPC` - 32-bit next program counter. Note that this value may not always be $pc+4$
+   when executing a branch/jump delay slot.
+1. `lo` - 32-bit MIPS LO special register.
+1. `hi` - 32-bit MIPS HI special register.
+1. `registers` - General-purpose MIPS32 registers. Each register is a 32-bit value.
+
+A thread is represented by packing the above fields, in order, into a 166-byte buffer.
+
+### Thread Hash
+
+A thread hash is computed by hashing the 166-byte thread state buffer with the Keccak256 hash function.
+
+### Thread Stack Hashing
+
+> **Note:** The `++` operation represents concatenation of 2 byte string arguments
+
+Each thread stack is represented in the FPVM state by a "hash onion" construction using the Keccak256 hash
+function. This construction provides a succinct commitment to the contents of a thread stack using a single `bytes32`
+value:
+
+- An empty stack is represented by the value:
+  - `c0 = hash(bytes32(0) ++ bytes32(0))`
+- To push a thread to the stack, hash the concatenation of the current stack commitment with the thread hash:
+  - `c1 = hash(c0 ++ hash(el0))`.
+- To push another thread:
+  - `c2 = hash(c1 ++ hash(el1))`.
+- To pop an element from the stack, peel back the last hash operation:
+  - `c3 = c1`
```

Comment:
it's not clear that we're popping `c2` in the `c3` example. Consider using a function notation for clarity:
```
push(c1, el1) => c2 = hash(c1 ++ hash(el1))
pop(c2) => c3 = c1
```